### PR TITLE
Issue122: Matching type question

### LINKED ISSUE #122: Question type matching

### LINKED ISSUE #129: Multiple choice is not using formatting correctly

### PR DESCRIPTION
**Pull request by esteeroramy**
fixes #122 
fixes #129 
Validation will come soon, currently blocked on #128 
matchingquestion

### CODE REVIEW

File: public/js/mc-answer.js [hunk 1]
```diff
@@ -6,7 +6,7 @@ var mcAnswerCount = 4; // default number of options
 var addMCAnswers = function(dom){
     mcAnswerCount++;
     var newdiv = document.createElement('p');
-    var inputdiv = '<input type="radio" name="radbutton" value="mcans{0}" id="mcans{1}" required/><label for="mcans{2}">Is Correct</label><input type="text" name="mcans{3}" placeholder="Enter Answer Here" required style="float:left;" class="form-control"/><a onclick="$(this).parent().remove()" class="btn-floating btn-tiny waves-effect waves-light red"><i class="tiny material-icons">close</i></a>';
-    newdiv.innerHTML = inputdiv.format(mcAnswerCount,mcAnswerCount,mcAnswerCount,mcAnswerCount);
+    var inputdiv = '<input type="radio" name="radbutton" value="mcans{0}" id="mcans{0}" required/><label for="mcans{0}">Is Correct</label><input type="text" name="mcans{0}" placeholder="Enter Answer Here" required style="float:left;" class="form-control"/><a onclick="$(this).parent().remove()" class="btn-floating btn-tiny waves-effect waves-light red"><i class="tiny material-icons">close</i></a>';
+    newdiv.innerHTML = inputdiv.format([mcAnswerCount]);
```
Comment (mehtadarshan1) [inline public/js/mc-answer.js]: When you are fixing your conflicts, remember to keep the develop branch changes.

File: server/questions.js [hunk 2]
```diff
@@ -196,7 +202,35 @@ exports.checkAnswer = function(questionId, user, answer, callback) {
 			return callback('error', null);
 		}
 
-		var value = (answer === question.answer);
+		var value = null;
+
+		if (question.type === common.questionTypes.MATCHING.value && answer) {
+		    const ansLeftSide = answer[0];
```
Comment (mehtadarshan1) [inline server/questions.js:208]: Good idea, I think we should keep the matching question checking part in its own function and you can reference it in here by using cases as we do for other types. That way this function is clean and we can add more checking for other question types using that approach

**Review by mehtadarshan1 — CHANGES_REQUESTED** [review]
Great work, Just a few minor changes for code base

File: public/js/matching-answer.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
+var matchAnswerCount = 3; // default number of options
+
+// Answering Variables
+var answeredId = 0;
+var answerList = [];
+var currentSelected = null;
+
+/**
+* Adds an option to the multiple choice
```
Comment (Mohammed-JS-Khan) [inline public/js/matching-answer.js]: adds an option to the matching? or multiple choice?

File: public/js/mc-answer.js (hunk 1 above)
Comment (esteeroramy) [reply public/js/mc-answer.js]: yeah no worries

File: server/questions.js (hunk 2 above)
Comment (esteeroramy) [reply server/questions.js:208]: sounds good

File: public/js/matching-answer.js (hunk 3 above)
Comment (esteeroramy) [reply public/js/matching-answer.js]: oopsies will fix

File: server/common.js [hunk 4]
```diff
@@ -41,4 +42,18 @@ const userTypes = Object.freeze({
     ADMIN     : 0,
     STUDENT   : 1
 });
-exports.userTypes = userTypes;
\ No newline at end of file
+exports.userTypes = userTypes;
+
+var randomize = function(data) {
```
Comment (ahmedqarmout2) [inline server/common.js]: ramdomizeList, incase we want other randomization functions

Comment (esteeroramy) [reply server/common.js]: good idea

File: public/js/matching-answer.js (hunk 3 above)
Comment (Mohammed-JS-Khan) [reply public/js/matching-answer.js]: lol nw, i just tripped out for a sec 😜

File: server/questions.js [hunk 5]
```diff
@@ -196,7 +202,35 @@ exports.checkAnswer = function(questionId, user, answer, callback) {
 			return callback('error', null);
 		}
 
-		var value = (answer === question.answer);
+		var value = null;
+
+		if (question.type === common.questionTypes.MATCHING.value && answer) {
```
Comment (ahmedqarmout2) [inline server/questions.js:207]: I dont like this. Its a huge addition to the checkanswer function for only one new type. We need a better way to check the answers based on type. I will let this one in, but I will refactor all of this later

File: server/questions.js [hunk 6]
```diff
@@ -196,7 +202,35 @@ exports.checkAnswer = function(questionId, user, answer, callback) {
...
+				const ansRightSide = answer[1];
+
+				if (ansLeftSide.length === question.leftSide.length) {
+					var checkIndexLeft;
+					var checkIndexRight;
+
+					for (i=0; i<ansLeftSide.length; i++) {
+					    checkIndexLeft = question.leftSide.indexOf(ansLeftSide[i]);
+							checkIndexRight = question.rightSide.indexOf(ansRightSide[i]);
+							console.log(checkIndexLeft);
```
Comment (ahmedqarmout2) [inline server/questions.js]: remove logs plz

**Review by Mohammed-JS-Khan — APPROVED** [review]
Awesome work fam! 🏋️

File: server/questions.js (hunk 6 above)
Comment (ahmedqarmout2) [reply server/questions.js]: not fixed

File: server/questions.js [hunk 7]
```diff
@@ -196,7 +202,35 @@ exports.checkAnswer = function(questionId, user, answer, callback) {
...
+							console.log(checkIndexRight);
```
Comment (ahmedqarmout2) [inline server/questions.js]: not fixed

Comment (esteeroramy) [reply server/questions.js]: I wasn't done

**Review by mehtadarshan1 — APPROVED** [review]

File: server/common.js [hunk 8]
```diff
@@ -41,4 +42,18 @@ const userTypes = Object.freeze({
...
     STUDENT   : 1
 });
-exports.userTypes = userTypes;
\ No newline at end of file
+exports.userTypes = userTypes;
+
+var randomizeList = function(data) {
+    var oldIndex, newIndex, tempHolder;
+
+    for (oldIndex=data.length-1; oldIndex > 0; oldIndex--) {
```
Comment (ahmedqarmout2) [inline server/common.js]: spaces

Comment (esteeroramy) [reply server/common.js]: sure

File: server/questions.js [hunk 9]
```diff
@@ -196,7 +202,33 @@ exports.checkAnswer = function(questionId, user, answer, callback) {
...
+
+		if (question.type === common.questionTypes.MATCHING.value && answer) {
+		    const ansLeftSide = answer[0];
+				const ansRightSide = answer[1];
+
+				if (ansLeftSide.length === question.leftSide.length) {
+					var checkIndexLeft;
+					var checkIndexRight;
+
+					for (i=0; i<ansLeftSide.length; i++) {
```
Comment (ahmedqarmout2) [inline server/questions.js]: spaces

Comment (esteeroramy) [reply server/questions.js]: okee

**Review by ahmedqarmout2 — APPROVED** [review]

**Pull request merged**